modules/renter/repair.go

```diff
@@ -26,323 +25,374 @@ var renewThreshold = func() types.BlockHeight {
+			for j := 0; j < file.erasureCode.NumPieces(); i++ {
+				potentialPieceGaps[i] = struct{}{}
+			}
+			potentialContractGaps := make(map[types.FileContractID]struct{})
+			for _, contract := range contracts {
+				potentialContractGaps[contract] = struct{}{}
 			}
 
-			// create contract entry, if necessary
-			f.mu.Lock()
-			contract, ok := f.contracts[host.ContractID()]
-			if !ok {
-				contract = fileContract{
-					ID:          host.ContractID(),
-					IP:          host.Address(),
-					WindowStart: host.EndHeight(),
-				}
+			// Delete every available piece from the potential piece gaps,
+			// and every utilized contract form the potential contract
+			// gaps.
+			for _, fcid := range contractMap[i] {
+				delete(potentialContractGaps, fcid)
+			}
+			for _, piece := range pieceMap[i] {
+				delete(potentialPieceGaps, piece)
 			}
 
-			// update contract
-			contract.Pieces = append(contract.Pieces, pieceData{
-				Chunk:      chunkIndex,
-				Piece:      pieceIndex,
-				MerkleRoot: root,
-			})
-			f.contracts[host.ContractID()] = contract
-			f.mu.Unlock()
-			errChan <- nil
-		}(missingPieces[i], hosts[i])
-	}
-	var errs hostErrs
-	for i := 0; i < numPieces; i++ {
-		err := <-errChan
-		if err != nil {
-			errs = append(errs, err)
-		}
-	}
-	if errs != nil {
-		return errs
-	}
+			// Merge the gaps into a slice.
+			var gaps chunkGaps
+			for fcid := range potentialContractGaps {
+				gaps.contracts = append(gaps.contracts, fcid)
+			}
+			for piece := range potentialPieceGaps {
+				gaps.pieces = append(gaps.pieces, piece)
+			}
 
-	return nil
-}
+			// Figure out the largest number of workers that could be
+			// repairing this piece simultaneously.
+			if len(potentialPieceGaps) < len(potentialContractGaps) {
+				gaps.numGaps = len(potentialPieceGaps)
+			} else {
+				gaps.numGaps = len(potentialContractGaps)
+			}
 
-// incompleteChunks returns a map of chunks containing pieces that have not
-// been uploaded.
-func (f *file) incompleteChunks() map[uint64][]uint64 {
-	f.mu.RLock()
-	defer f.mu.RUnlock()
+			// Record the number of gaps that this chunk has, which makes
+			// blocking-related decisions easier.
+			gapCounts[gaps.numGaps] = gapCounts[gaps.numGaps] + 1
 
-	present := make([][]bool, f.numChunks())
-	for i := range present {
-		present[i] = make([]bool, f.erasureCode.NumPieces())
-	}
-	for _, fc := range f.contracts {
-		for _, p := range fc.Pieces {
-			present[p.Chunk][p.Piece] = true
-		}
-	}
+			// Create a name for the incomplete chunk.
+			chunkPrefix := make([]byte, 8)
+			binary.LittleEndian.PutUint64(chunkPrefix, i)
+			chunkName := string(chunkPrefix) + file.name
 
-	incomplete := make(map[uint64][]uint64)
-	for chunkIndex, pieceBools := range present {
-		for pieceIndex, ok := range pieceBools {
-			if !ok {
-				incomplete[uint64(chunkIndex)] = append(incomplete[uint64(chunkIndex)], uint64(pieceIndex))
-			}
+			// Add the chunk to the repair matrix.
+			repairMatrix[chunkName] = &gaps
```

Comment:
why use a pointer? `chunkGaps` is just a wrapper around two pointers (and an int, I guess)